src/Clients/GuzzleClient.php

```diff
@@ -18,47 +20,99 @@ class GuzzleClient implements Client
      */
     const DEFAULT_METHOD = 'GET';
 
+    /**
+     * Supported HTTP methods
+     */
+    const SUPPORTED_METHODS = [
+        'GET' => true,
+        'HEAD' => true,
+        'POST' => true,
+        'PUT' => true,
+        'DELETE' => true,
+        'OPTIONS' => true,
+    ];
+
     /**
      * @var array the Client main options
      */
     private $mainOptions;
 
+
     public function __construct(array $mainOptions = [])
     {
         $this->mainOptions = $mainOptions;
     }
 
     /**
      * {@inheritdoc}
+     * @throws UnavailableServiceException
      */
     public function request($resource, array $options)
     {
         try {
-            $client = new OriginalGuzzleClient($this->mainOptions);
-            $method = $this->defineMethod($options);
-            $options['http_errors'] = true;
+            $client = $this->buildClient();
+            $method = $this->getHttpMethod($options);
+            $options['exceptions'] = true;
+
+            // prevents unhandled method errors in Guzzle 5
+            unset($options['method']);
 
-            return (string) $client->request($method, $resource, $options)->getBody();
-        } catch (Exception $exception) {
-            throw new UnavailableService($exception->getMessage());
+            $request = $client->createRequest($method, $resource, $options);
+            return (string) $client->send($request)->getBody();
+        } catch (Exception $e) {
+            throw new UnavailableServiceException($e->getMessage(), $e->getCode(), $e);
         }
     }
 
     /**
      * @param array $options the list of options
      *
      * @return string the method
+     * @throws UnsupportedMethodException
      */
-    private function defineMethod(array $options)
+    private function getHttpMethod(array $options)
     {
         if (isset($this->mainOptions['method'])) {
             return $this->mainOptions['method'];
         }
 
         if (isset($options['method'])) {
+            if (!array_key_exists($options['method'], self::SUPPORTED_METHODS)
+                || !self::SUPPORTED_METHODS[$options['method']]
+            ) {
+                throw UnsupportedMethodException::unsupportedMethod($options['method']);
+            }
             return $options['method'];
         }
 
         return self::DEFAULT_METHOD;
     }
+
+    /**
+     * @return OriginalGuzzleClient
+     */
+    private function buildClient()
+    {
+        if (isset($this->mainOptions['mock']) && $this->mainOptions['mock'] instanceof Mock) {
+            return $this->buildMockClient($this->mainOptions['mock']);
+        }
+
+        return new OriginalGuzzleClient($this->mainOptions);
+    }
+
+    /**
+     * Builds a client with a mock
+     * @return OriginalGuzzleClient
+     */
+    private function buildMockClient(Mock $mock)
```

Comment:
you made the class responsible of its mock: wtf?

Reply:
No, it's responsible for allowing you to mock responses in its client. The alternative was either exposing the Guzzle client via getter, or injecting the client in the constructor, which would require major modifications in the class design because it creates a new client with each new request.

In any case, this whole commit is supposed to be reverted once we are able to upgrade Guzzle again.

Reply:
Yes, really weird to add mock here :open_mouth:

Reply:
How about creating a new class named Guzzle5Client instead? My classes are final for a reason: they are fully tested and have no reason to change.

If you need different behavior, you should create your own concrete class: it's not a big job, you can copy-paste the code.

My only worry is how we can test the 2 clients while they depend on 2 major versions of Guzzle?

Reply:
Solved. Ignore this comment.

Reply:
We cannot test them at the same time (only one can be installed at a given time), but I guess it could be made to support both by probing the existing classes... I guess?

Anyway this could be done later on.